Increase the rule speed

workgroupengineering opened the pull request:
### This is preliminary release for testing purpose. ###
Here I implemented my # 1771 proposal to improve the performance of the rules.
I did not test on Esp32 because i haven't it.
Can you suggest an esp32 card to buy that supports CAM BUS and BLE?

workgroupengineering committed:
Moved the rules related functions from Misc.ino to ESPEasyRules.ino

workgroupengineering committed:
implement matching of rule file

workgroupengineering committed:
Added addButton, addFormTextBox overload and  addSaveButton, AddDeleteButton functions

workgroupengineering committed:
Redefine Rules WebUi

workgroupengineering committed:
Change menu url from relative to absolute

workgroupengineering committed:
Fixed ESP32 build

TD-er commented:
All ESP32 boards do have support for BLE.
Not sure what you mean by "CAM BUS".
Do you mean "CAN BUS", like used in cars? Or the camera module that some dev boards seem to support?

TD-er commented:
Can you also explain a bit what is included in this PR?
It looks like a lot more has changed than only the execution speed of rules.
Please also give some indication where speed improvements were made. I noticed a part where you check if rules are enabled, but I guess there is more :)

TD-er commented:
Does this PR conflict with this one?
https://github.com/letscontrolit/ESPEasy/pull/1805

workgroupengineering commented:
> 
> 
> All ESP32 boards do have support for BLE.
> Not sure what you mean by "CAM BUS".
> Do you mean "CAN BUS", like used in cars? Or the camera module that some dev boards seem to support?
it's my mistake, I wanted to say the can bus.


> 
> 
> Can you also explain a bit what is included in this PR?
> It looks like a lot more has changed than only the execution speed of rules.
> Please also give some indication where speed improvements were made. I noticed a part where you check if rules are enabled, but I guess there is more :)


The core of PR is: 

https://github.com/workgroupengineering/ESPEasy/blob/96c380152a09a852dafb40ce0b2fa4f13e478751/src/Rules.ino#L1-L57

This code map event name to the file and passed it to rulesProcessingFile.
All other is only for Web UI of rule.
For the speed texts I have not yet done de banckmark to be shared, but from the tests I have done there is a marked improvement when the rules file was very large.


> 
> 
> Does this PR conflict with this one?
> #1805



> 
> 
> Does this PR conflict with this one?
> #1805

When merge the PR #1805 i rebase this PR on it.

workgroupengineering commented:
@TD-er Can you explain how function START_TIMER and STOP_TIMER?

uzi18 commented on src/WebServer.ino:
```diff
@@ -477,7 +477,18 @@ void WebServerInit()
+  WebServer.on(F("/rules/"), []()
+  {
+    WebServer.sendHeader(F("Location"), F("/rules"),true);
+    WebServer.send(302, F("text/plane"),F(""));
```
text/plain ??

TD-er commented:
@workgroupengineering 
They are macro's.

```c++
#define START_TIMER const unsigned statisticsTimerStart(micros());
#define STOP_TIMER_TASK(T,F)  if (mustLogFunction(F)) pluginStats[T*32 + F].add(usecPassedSince(statisticsTimerStart));
#define STOP_TIMER_LOADFILE miscStats[LOADFILE_STATS].add(usecPassedSince(statisticsTimerStart));
#define STOP_TIMER(L)       miscStats[L].add(usecPassedSince(statisticsTimerStart));
```

`START_TIMER` does declare a variable, so it must be in the same scope as where `STOP_TIMER` is called.

` STOP_TIMER` is called with an integer value.
You have to define that integer and also make some descriptive text for it. For example take a look at `SEND_DATA_STATS`.

`STOP_TIMER_TASK` is specific to timing of a task (plugin), where `T` is the Task and `F` is the function of that task. (switch statement in the task `.ino` file)

workgroupengineering commented:
@uzi18 
> 
> 
> text/plain ??

thank you, It is typo, i will fixing it when io do rebase.


@TD-er 
> 
> 
> @workgroupengineering
> They are macro's.
> 
> ```c++
> #define START_TIMER const unsigned statisticsTimerStart(micros());
> #define STOP_TIMER_TASK(T,F)  if (mustLogFunction(F)) pluginStats[T*32 + F].add(usecPassedSince(statisticsTimerStart));
> #define STOP_TIMER_LOADFILE miscStats[LOADFILE_STATS].add(usecPassedSince(statisticsTimerStart));
> #define STOP_TIMER(L)       miscStats[L].add(usecPassedSince(statisticsTimerStart));
> ```
> 
> `START_TIMER` does declare a variable, so it must be in the same scope as where `STOP_TIMER` is called.
> 
> `STOP_TIMER` is called with an integer value.
> You have to define that integer and also make some descriptive text for it. For example take a look at `SEND_DATA_STATS`.
> 
> `STOP_TIMER_TASK` is specific to timing of a task (plugin), where `T` is the Task and `F` is the function of that task. (switch statement in the task `.ino` file)

thank you ,is there header file with all used constant?

TD-er commented:
If you search for `SEND_DATA_STATS` you will find all (2) places where you have to add a new one and where the others are defined.

TD-er commented:
@workgroupengineering 
Can you have a look at these merge conflicts?
Or should I?
Last few days we have had a number of "issues" in the rules processing department. See https://github.com/letscontrolit/ESPEasy/issues/1818

workgroupengineering commented:
I resolved merge conflict.

TD-er commented:
I've looked at it on my test node and this can indeed be a big improvement in rules speed and usability.

But, since this is really a big change, I think some users should test it.
@Grovkillen since you're the rules guy, please make some tests with this.
And tip: backup your rules, since this redesign will split the rules in separate files per rule.

Also I don't know what the filesystem overhead is for small files. For example if it is 4k, then it will only allow about 16 rules on a system with 1MB flash.

TD-er commented:
To be honest, I am not really sure what to do with this PR.
It is breaking existing installations, since the existing rules cannot be used. They are not split into other files for example as a transition.

But it really has some nice improvements and I would love to see them being used.
One of the things is to move the rules functions to a separate file, to make it more comprehensible.
But it will break other PRs.

I think I will break-up this PR and use bits of it in little steps and we have to think about splitting the rules file.
It will also solve the issue of large POST requests and make the parsing a lot less memory intensive.

Anybody else some suggestions here?

Grovkillen commented:
Okay I've tested the code and I also like some of the new ways of thinking but it's a huge switch of how it's been handled before and I'm also a bit perplexed on how to go about this.

giig1967g commented:
I haven't looked into the code, but would it be possible to maintain both versions of the rule engine with an option to select which version a user want to use?
So if a user wants to use old version, it will keep exixsting structure, otherwise, if a user wants the new version will explicitally set this mode and use it.

Also, @Grovkillen did you checked if there are limitations on the number of rules as per @TD-er warning?

Grovkillen commented:
I've tried it but I cannot get timer events and custom events to trigger.

Grovkillen commented:
@giig1967g haven't pushed the size yet

Grovkillen commented:
But to conclude, I think we will use some of these ideas in future releases! :)

Grovkillen commented:
![2018-10-09 10 42 53](https://user-images.githubusercontent.com/26706274/46657377-6db41080-cbb0-11e8-8944-15ca0fc704d1.png)
![2018-10-09 10 43 28](https://user-images.githubusercontent.com/26706274/46657378-6db41080-cbb0-11e8-99c4-cd995d1c11b6.png)

workgroupengineering commented:
the core of this PR and in the following functions:
```c++
String EventToFileName(String& eventName)
{
  int size = eventName.length();
  int index = eventName.indexOf('=');
  if(index >- 1)
  {
    size = index;
  }
#if defined(ESP8266)
  String fileName = F("rules/");
#endif
#if defined(ESP32)
  String fileName = F("/rules/");
#endif
  fileName += eventName.substring(0,size);
  fileName.replace('#',RULE_FILE_SEPARAROR);
  fileName.toLowerCase();
return fileName;
}


String FileNameToEvent(String& fileName)
{
  #if defined(ESP8266)
  String eventName = fileName.substring(6);
  #endif
  #if defined(ESP32)
  String eventName = fileName.substring(7);
  #endif
  eventName.replace(RULE_FILE_SEPARAROR,'#');
  return eventName;
}

void rulesProcessing(String& event)
{
  checkRAM(F("rulesProcessing"));
  unsigned long timer = millis();
  if (loglevelActiveFor(LOG_LEVEL_INFO)) {
    String log = F("EVENT: ");
    log += event;
    addLog(LOG_LEVEL_INFO, log);
  }


  String fileName = EventToFileName(event);
  //if exists processed the rule file
  if(SPIFFS.exists(fileName))
      rulesProcessingFile(fileName, event);
  else
    addLog(LOG_LEVEL_DEBUG
      , String(F("EVENT: ")) + event
        + String(F(" is ingnored. File ")) + fileName
        + String(F(" not found.")));


  if (loglevelActiveFor(LOG_LEVEL_DEBUG)) {
    String log = F("EVENT: ");
    log += event;
    log += F(" Processing time:");
    log += timePassedSince(timer);
    log += F(" milliSeconds");
    addLog(LOG_LEVEL_DEBUG, log);
  }
  backgroundtasks();


}
```

everything else is for the user interface.

I weigh that the two solutions can coexist as suggested by @giig1967g . However, it is also possible to perform a function which, after updating the firmware, detects whether the Rules1.txt,..... files exist, and divides them according to the events.

TD-er commented:
Is there a check on the file name length?
I just read a filename cannot exceed 32 characters.
That may explain why @Grovkillen 's tests failed?

Grovkillen commented:
Just to make sure, the screenshot contains error in the syntax of the set timer command ;)

Grovkillen commented:
Okay, adding `on ... do` made the rules kick in. Now that is not ideal. I was under the impression that these chunks of files would replace the on sections.

Grovkillen commented:
Because I figured this way we could have a better distinction between event based triggers and script like execution of code.

workgroupengineering commented:
> Is there a check on the file name length?
> I just read a filename cannot exceed 32 characters.

it is checked on UI.




> That may explain why @Grovkillen 's tests failed?

Missing on ... do

I kept it first of all for compatibility with past and second because you can use multiple rules of the same family in the same file as:


```
On Rules#Timer=1 do  //When Timer1 expires, do
   servo,1,12,30
   timerSet,2,1       //Set Timer 2 for the next event in 1 second
 endon
 On Rules#Timer=2 do  //When Timer2 expires, do
   servo,1,12,0
   timerSet,1,30      //Set Timer1 for the next event in 30 seconds
 endon

```

workgroupengineering committed:
Adding flag to settings for switch from old to new rules engine (Default engine old).

workgroupengineering commented:
I have adding flag to settings for switch from old to new rules engine, default value i old engine.

Grovkillen commented:
I have tested the flag and both rule engines seems to coexist just fine. I like it this way because then we can do some experimental stuff on the new rule engine.

Maybe we should have the name in the advanced page stating "old rule engine" instead of "monolithic engine" 😄 Even though I can agree that it's monolithic.

Grovkillen commented:
But we really should remove the need for `on...endon` for the new engine. And the timer events should maybe be structured with 1...4 input fields, one per timer?

workgroupengineering commented:
I have rename MonolithicRuleEngine flag.

workgroupengineering commented:
> But we really should remove the need for `on...endon` for the new engine. And the timer events should maybe be structured with 1...4 input fields, one per timer?

I do not know what to answer, I have to think about it.

@TD-er  what do you think about it?

Grovkillen commented:
Since we want the new engine to be event based we need to make sure that we're thinking dynamically for the future.

TD-er commented:
@workgroupengineering About the `on ...` .... `endon...` part.
Maybe you can add these as some kind of template around the rules text field, to indicate they are not needed anymore but are considered to be the scope.

Did you also have a look at the minimum space a file takes on SPIFFS?
The documentation is not really clear about this, but I have seen some mention about 4kB.
So maybe we can also have some intermediate way of file handling.
I guess on average a on..endon scope is often less than 1 kB, so maybe we can combine a few of these in the same file.

The free SPIFFS size is given on the sysinfo page, so maybe we can see jumps of 4kB when adding a file containing 1 byte.

TD-er commented:
What is the current status of this PR?

TD-er committed:
[Rules engine] Set UseOldRulesEngine as default

Changed to use one of the VariousBits in the settingsfile, not adding an extra bit.

workgroupengineering commented:
I have little time to carry it forward. What little I had I used to do the rebase.

TD-er commented:
I am now looking into it, to see what impact it has and what else needs to be done.
So don't consider it to be time lost :)

TD-er commented:
I just tested a minimal build, to see if this one increases the binary a lot.
And it does, it is now 11k larger, so the minimal build cannot be used for OTA anymore.
Any idea what may be the biggest increase in size?
You did add a few (small) SVG's, but I don't think that's the real big part of the increase.

workgroupengineering commented:
Can depend from this b14fe5de3224aec81de12f4a9dca7d997b3fe1bd?

TD-er committed:
[Stack] Reduce allocations on stack

TD-er committed:
[Stack] Make IP-address formatting simpler

TD-er committed:
[Stack] Optimize GetArgv to only allocate what is needed.

And also only allocate on the heap, not on the stack.

TD-er commented:
I made some more updates to reduce the stack.
Still have to fix some calls to make the IR builds work, but for those who want to test.
[New test build](https://www.dropbox.com/s/q74w3fd7p406ncs/ESPEasy_mega-20181201-12-gdeab18a.zip?dl=0)

In my tests, the stack peak usage was about 200 bytes better. (200 bytes more free)

TD-er committed:
[Stack] P035_IRTX reduction of memory usage & speed improvement

- Parsed commands or IR codes now only use amount of memory needed, no fixed allocation on the stack.
- Strings passed to functions via const reference instead of deepcopy
- Massive checks for brand name is now done on lower case, to improve speed.

TD-er committed:
[Uncrustify] Make rules source code files better readable

Since the `ESPEasyRules.ino` file is newly introduced in this branch, now is the best moment to tidy up the source code.
Also included the used Uncrustify.cfg file.

TD-er committed:
[Memory] Pass String as reference or const reference, not deepcopy

A number of functions was given a String parameter as a deepcopy instead of a (const) reference.
This will use a lot more memory and time, where it isn't needed.
Also those deepcopy actions may fail silently and thus lead to crashes.

TD-er commented:
And a new [test build](https://www.dropbox.com/s/2z0lugi3jkjo01v/ESPEasy_mega-20181201-15-ga784a43.zip?dl=0)
This one is including this commit: https://github.com/letscontrolit/ESPEasy/pull/1813/commits/aaae79c8d90fb38084da494c2495c48d608d5231

Free RAM has increased, free stack has increased and the lowest free stack value is about 400 bytes better now.

Please test this one thoroughly, since all getArgv functions have changed and also a lot of stack allocated stuff has been removed.

TD-er committed:
[Rules] GetArgv should clear the return string before filling it

The `GetArgv` function should clear the return string, since it is re-used in `parseCommandString`

See: https://github.com/letscontrolit/ESPEasy/issues/2112#issuecomment-443494191

TD-er commented:
[New test build](https://www.dropbox.com/s/rbsu8jaexj5gera/ESPEasy_mega-20181201-16-g62cc792.zip?dl=0)

TD-er committed:
[ESP32] Disable RTOS Multitasking option (multi_heap_poisoning)

When enabled, we now get crashes and CORRUPT HEAP messages.

TD-er committed:
[PWM] #2117  PWM stops after about 80 msec in core 2.4.1

See #2117

TD-er committed:
[webserver CORS] #2119 Allow commands to be scripted

See #2119

TD-er commented:
No really big breaking issues found in the test builds?

TD-er committed:
[HLW8012] Update library to 1.1.1

TD-er committed:
[HLW8012] Make reading Sonoff POW non blocking ( #2097 )

See #2097
Timings are now less than a msec per iteration and interrupt driven.
Also settings are no longer saved every time the plugin gets initialized. (flash wear out)

giig1967g commented:
@workgroupengineering 
Hi, can you give us a tutorial on how to use the new rules engine?
For example: 
a) how can I edit the rule event name?
b) how can I create the following events: task#value=1 and task#value=0

Enhancements:
1) would it be possible to expand/collapse the event items in the rules page so it could be possible to see and even edit the rules in a single page?
2) would it be possible to reorder the rules or sort them in the list?

workgroupengineering commented:
> a) how can I edit the rule event name?
> b) how can I create the following events: task#value=1 and task#value=0

if i have some time this weekend. I'll do it.




>Enhancements:
>
 >  1.  would it be possible to expand/collapse the event items in the rules page so it could be possible to see and even edit the rules in a single page?
>  2.   would it be possible to reorder the rules or sort them in the list?

Great ideas, but I don't have time to do it right now.

uzi18 commented:
maybe we can track available file names in rules dir, this will kick even more, what do you think?

TD-er commented:
We still have to look into the allocation size of a single small file.
It is very likely to be 4k.

So maybe we should use some metadata list to keep track of rules in a bigger file.

uzi18 commented:
couldn't we change it to 512 bytes?

TD-er commented:
Not sure and not sure how.

giig1967g commented:
> 
> 
> We still have to look into the allocation size of a single small file.
> It is very likely to be 4k.
> 
> So maybe we should use some metadata list to keep track of rules in a bigger file.

But in this case we can use the old rules.txt file with the metadata, couldn't we?

TD-er commented:
@giig1967g Yes we can.
But the "new" situation is meant to be "testing" for now and should lead to a single solution.

This development phase should lead to good insights on how it will be used and based on those experiences we must make decision on how the rules will be stored.
Maybe we can edit, working on a number of small files or larger files with a lot of empty space and then generate a single (compact) file with index.
Somewhat comparable with "defrag" on a desktop PC.

uzi18 commented:
does ` on event do endon` combo is still needed in new way per file?

TD-er commented:
> does ` on event do endon` combo is still needed in new way per file?

At this moment it looks like it is still needed.
Do we have any other way to indicate a new rule block?

uzi18 commented:
my question is because i'm thinking about  some kind of indexer for event per file...

this could improve even old rules

TD-er commented:
@uzi18 I think that would be great to have.
And personally I think we should still keep working on larger files, but allow the option to edit only a single rule part.
In the back, when saving it, we should add the new one to a file (append?) and update the index.
At some point this old part could be removed, but as long as the index does not point to the old version, it will not be used.
This allows for quite some complex rule situations, where you can even temporarily enable/disable rules or switch between rules at runtime.

uzi18 commented:
my idea is simple, will try to prepare some kind of PR and next we can think if we want to extend this concept, but it will be compatible with both rules

uzi18 commented:
my idea is simple, will try to prepare some kind of PR and next we can think if we want to extend this concept, but it will be compatible with both rules

uzi18 commented:
my idea is simple, will try to prepare some kind of PR and next we can think if we want to extend this concept, but it will be compatible with both rules

uzi18 commented:
my idea is simple, will try to prepare some kind of PR and next we can think if we want to extend this concept, but it will be compatible with both rules
